src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,473 @@
+                    assert len(batch) == len(outputs), f"result has {len(outputs)} items but batch is {len(batch)}"
+                    result = {request[0]: r for request, r in zip(batch, outputs)}
+                    self._responses.update(result)
+        except Exception as e:
+            result = {request[0]: e for request in batch}
+            self._responses.update(result)
+
+    async def consumer(self):
+        while True:
+            await asyncio.sleep(0.05)
+
+            has_sent = False
+
+            batch = self._batch[: self.max_batch_size]
+            while batch and (
+                (len(batch) >= self.max_batch_size) or ((time.time() - self._last_batch_sent) > self.timeout_batch)
+            ):
+                has_sent = True
+
+                asyncio.create_task(self.send_batch(batch))
+
+                self._batch = self._batch[self.max_batch_size :]
+                batch = self._batch[: self.max_batch_size]
+
+            if has_sent:
+                self._last_batch_sent = time.time()
+
+    async def process_request(self, data: BaseModel):
+        if not self.servers:
+            raise HTTPException(500, "None of the workers are healthy!")
+
+        request_id = uuid.uuid4().hex
+        request: Tuple = (request_id, data)
+        self._batch.append(request)
+
+        while True:
+            await asyncio.sleep(0.05)
+
+            if request_id in self._responses:
+                result = self._responses[request_id]
+                del self._responses[request_id]
+                _raise_granular_exception(result)
+                return result
+
+    def run(self):
+        if self._server_ready:
+            return
+
+        INPUT_SCHEMA = self._input_schema
+        OUTPUT_SCHEMA = self._output_schema
+
+        logger.info(f"servers: {self.servers}")
+
+        self._ITER = cycle(self.servers)
+        self._last_batch_sent = time.time()
+
+        fastapi_app = _create_fastapi("Load Balancer")
+        fastapi_app.global_request_count = 0
+        fastapi_app.num_current_requests = 0
+        fastapi_app.last_process_time = 0
+        fastapi_app.SEND_TASK = None
+
+        @fastapi_app.on_event("startup")
+        async def startup_event():
+            fastapi_app.SEND_TASK = asyncio.create_task(self.consumer())
+            self._server_ready = True
+
+        @fastapi_app.on_event("shutdown")
+        def shutdown_event():
+            fastapi_app.SEND_TASK.cancel()
+            self._server_ready = False
+
+        @fastapi_app.get("/system/info", response_model=_SysInfo)
+        async def sys_info():
+            return _SysInfo(
+                num_workers=len(self.servers),
+                servers=self.servers,
+                num_requests=fastapi_app.num_current_requests,
+                process_time=fastapi_app.last_process_time,
+                global_request_count=fastapi_app.global_request_count,
+            )
+
+        @fastapi_app.put("/system/update-servers")
```

Comment:
Wouldn't malicious API consumers be able update servers also? We should protect this.

Reply:
we can add the authentication from Muse - https://github.com/Lightning-AI/stable-diffusion-deploy/blob/main/muse/components/load_balancer.py#L216

Reply:
yes, that would be ok